### PR TITLE
micropkg packaging improvements

### PR DESCRIPTION
**Pull request by astrojuanlu**
## Description
- Fix gh-2567.
- Fix gh-2542.
- Beginning of gh-2414 (still generates a `setup.py` on the fly, but the code is ready to use a `pyproject.toml` instead).
- Obliterate any remaining usage of `pkg_resources`, which is deprecated and will go away any time.

## Development notes
All the packaging stuff is quite complex - in fact, after an initially good streak, this is now my second backtrack attempt, after opening gh-2569 and gh-2597 (hopefully those two can go next). Keeping the micropackaging workflow intact while removing the strong assumptions around `setup.py` was extra hard.

I even decided to have a runtime dependency on `setuptools` because I couldn't find any "blessed", standalone PyPA package to do some tasks.

It's been mentioned that this workflow is weird and that we should evaluate whether we want to keep it or not https://github.com/kedro-org/kedro/milestone/21 but this work blocks the rest of the progress in https://github.com/kedro-org/kedro/milestone/36, so I didn't want to block it on that decision (which could be taken after 0.19).

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

### CODE REVIEW

**Comment by astrojuanlu** [comment]
A small number of corner cases is failing, will have a look at those later this week.

**Comment by astrojuanlu** [comment]
The last set of test failures is due to a behavior change from `pkg_resources` to `packaging`, noted it here https://github.com/pypa/packaging/issues/644#issuecomment-1567982812

I am hoping that we get an official answer soon, but while we wait, I think this is ready for review.

**Review by stichbury — APPROVED** [review]
Happy with this from docs perspective, but obvs not the main point of the PR.

File: kedro/framework/cli/micropkg.py [hunk 1]
```diff
@@ -854,19 +912,36 @@ def _append_package_reqs(
     )
 
 
+def _get_all_library_reqs(metadata):
+    # Inspired by https://discuss.python.org/t/\
```
Comment (merelcht) [inline kedro/framework/cli/micropkg.py:916]: (nit) I personally find it more useful to have a short explanation of what the function does instead of just the link.

File: kedro/framework/cli/micropkg.py [hunk 2]
```diff
@@ -854,19 +912,36 @@ def _append_package_reqs(
...
     """
     parseable_requirements = set()
-    for requirement in pkg_resources.yield_lines(requirements):
-        try:
-            parseable_requirements.add(pkg_resources.Requirement.parse(requirement))
-        except ValueError:
-            continue
+    if isinstance(requirements, str):
+        requirements = requirements.splitlines()
+    # TODO: Properly handle continuation lines,
```
Comment (merelcht) [inline kedro/framework/cli/micropkg.py:935]: Do you still need to do this or is this blocked by something?

**Comment by merelcht** [comment]
I've tried packaging and pulling with this new version and the old (current) one and it all seems to work fine and to be backwards compatible 👌 It seems like the micropackage that's generated is still exactly the same. I guess that's supposed to be the case right?

**Comment by astrojuanlu** [comment]
Yes, absolutely! That's why this PR would not fully close gh-2414 - the generated micropkg pretty much still uses `setup.py`.

**Review by merelcht — APPROVED** [review]
This looks like to be working as expected, so I'm happy to have it merged in. Maybe add to the release notes which changes were made here for traceability? And thanks for helping to improve this part of messy code base @astrojuanlu 🙇 ⭐

File: kedro/framework/cli/micropkg.py (hunk 1 above)
Comment (astrojuanlu) [reply kedro/framework/cli/micropkg.py:916]: Good point, I added a proper docstring

**Comment by astrojuanlu** [comment]
I have addressed all the outstanding comments and now I believe all tests should pass. If everything goes well and unless there are other opinions, I'll merge this.

**Comment by astrojuanlu** [comment]
All tests passing, CI failures due to non-100 % coverage. Will need to add tests to cover the extra possible (but unlikely) error messages.

**Comment by astrojuanlu** [comment]
Added tests, finally! This is ready to go in. It has 2 approvals already but waiting until EOD to merge.

**Comment by astrojuanlu** [comment]
Meh, playing whack-a-mole. I can reproduce these new test failures locally, will tackle them later.

**Comment by merelcht** [comment]
> Added tests, finally! This is ready to go in. It has 2 approvals already but waiting until EOD to merge.

Ideally another engineer would approve this since the majority of the changes are in code.

File: kedro/framework/cli/micropkg.py [hunk 3]
```diff
@@ -47,6 +50,50 @@
 """
 
 
+class _EquivalentRequirement(Requirement):
```
Comment (SajidAlamQB) [inline kedro/framework/cli/micropkg.py:53]: Maybe we should be adding some docstrings here to describe the purpose of the class and its methods.

**Review by SajidAlamQB — APPROVED** [review]
Overall, the changes you've made to the micro-packaging workflow seem comprehensive and well thought out. Just some minor comments about readability. 

Thank you @astrojuanlu

File: kedro/framework/cli/micropkg.py (hunk 3 above)
Comment (astrojuanlu) [reply kedro/framework/cli/micropkg.py:53]: I'll add a docstring to make it more clear that this essentially overrides `packaging.requirements.Requirement.__eq__` to be backwards compatible with `pkg_resources`
